doc/PROTOCOL-SEMANTICS.md

```diff
@@ -0,0 +1,268 @@
+Service is a core distinguishing feature of its Methods; if two Methods have the
+same name but exist in different Services they must be considered distinct and
+not be confused. A Method name including its Service namespace prefix with a "/"
+separator is a "full method name".
+
+## Calls
+
+RPCs, or "Calls," are initiated by a client to a server, typically via a
+Channel. There may be multiple servers that _could_ have received the Call (as
+is common for load balancing), but only a single server may process an
+individual Call. Calls are assumed to be non-idempotent and may not be
+"replayed" except for when gRPC is explicitly informed it is safe to do so.
+
+Calls are natively two independent streams (i.e., full duplex bidirectional) of
+Messages. The request stream is started with Request Headers and ended by Half
+Close. The response stream is started with Response Headers and ended by
+Trailers, or consistes only of Trailers. Messages may exist between the headers
+and the end of the stream. Request Headers, Response Headers, Messages, Half
+Close, and Trailers are the units of communication and, absent the Call's
+termination, will be communicated to the remote without the need to send further
+units on the stream. However, see the optimizations permitted for unary Calls
+below.
+
+Request Headers contain the Full Method Name and Metadata. Response Headers
+contain Metadata. Trailers contain the Status and Metadata. Messages contain the
+Message Payload. These contents are not exhaustive; gRPC features may extend
+these concepts. It is quite common for features to add additional fields to
+Request Headers, Response Headers, and Trailers.
+
+The Call initiation is with Request Headers, within which the client
+indicates the method to be run by its Full Method Name. The Call is gracefully
+completed when the server responds with Trailers, which contains a Status
+communicating the success or failure of the RPC. If a server responds with
+Trailers before receiving the client's Half Close, then any unprocessed
+client-sent Messages and Half Close is lost. Note that on the server there is a
+period of time between when the server application responds with a Trailers and
+when that Trailers is actually sent; the Call is only truly complete when the
+Trailers is sent. Similarly, on the client there is a period between the gRPC
+implementation receiving the Trailers and when the application receives the
+Trailers; the Call is only truly complete when the Trailers is received by the
+application.
+
+Calls may terminate early by being "cancelled." Implementations must allow
+clients to cancel Calls, but cancellations may occur in other ways like I/O
+failures. A cancellation appears as a Trailers with a Status Code of CANCELLED
+to clients and is a special state on servers. Cancellation is an abrupt killing
+of the Call; inbound and outbound buffered data should be cleared. Cancellation
```

Comment:
Does this include headers?  If a server sends headers, and then error-trailers (or RST), should the client still process the initial headers?  (assuming it has not yet read them)  cc: @muxi

Reply:
Error-trailers aren't cancellation, so they should be in-line and the client should process the server-sent headers. RST is cancellation, but server doesn't really initiate cancellation today (I know it is possible in C, but that's another story). The only time a server cancels today is due to deadline, in which case the client should have already cancelled as well.

@carl-mastrangelo, why do you ask? What context is this coming up?

I'm noticing that this section should probably remove "I/O failures" from ways of cancellation, as we don't do abrupt killing of the call in that case.

Reply:
@muxi asked if server-headers should be dropped by the client, if the call is cancelled before they can be delivered to the application.   In Java I think we do deliver the complete data frames, but C-core does not, hence the question.

Reply:
Talked with @muxi offline, and the cancellation being discussed was client-initiated. In that case, yes, the server-sent headers should be dropped (excepting races).